pkg/testing/fixture_install.go

```diff
@@ -193,98 +206,127 @@ func (f *Fixture) installNoPkgManager(ctx context.Context, installOpts *InstallO
+		f.cleanUpCollectDiagnostics()
 
-			filePath := filepath.Join(dir, "build", "diagnostics", fmt.Sprintf("TEST-%s-%s-%s-ProcessDump.json", sanitizedTestName, f.operatingSystem, f.architecture))
-			f.t.Logf("Dumping running processes in %s", filePath)
-			file, err := os.OpenFile(filePath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
-			if err != nil {
-				f.t.Logf("failed to dump process; failed to create output file %s root: %s", file.Name(), err)
-				return
-			}
-			defer func(file *os.File) {
-				err := file.Close()
-				if err != nil {
-					f.t.Logf("error closing file %s: %s", file.Name(), err)
-				}
-			}(file)
-			err = json.NewEncoder(file).Encode(procs)
-			if err != nil {
-				f.t.Logf("error serializing processes: %s", err)
-			}
+		if f.cleanUpKeepInstalled() {
+			return
 		}
-	})
 
-	f.t.Cleanup(func() {
+		if f.cleanUpUninstall() {
+			return
+		}
+
+		f.dumpProcessOnFailure()
+
 		// check for running agents after uninstall had a chance to run
 		assert.Empty(f.t, getElasticAgentProcesses(f.t), "there should be no running agent at the end of the test")
 	})
 
-	f.t.Cleanup(func() {
-		f.t.Logf("[test %s] Inside fixture cleanup function", f.t.Name())
-
-		if !f.installed {
-			f.t.Logf("skipping uninstall; agent not installed (fixture.installed is false)")
-			// not installed; no need to clean up or collect diagnostics
-			return
+	emptyEnrollOpts := EnrollOpts{}
+	if enrollOnlyOpts != emptyEnrollOpts {
+		enrollArgs := []string{"enroll"}
+		enrollArgs = append(enrollArgs, enrollOnlyOpts.toCmdArgs()...)
+		out, err = f.Exec(ctx, enrollArgs, opts...)
+		if err != nil {
+			return out, fmt.Errorf("error running agent enroll command: %w", err)
 		}
+	}
 
-		// diagnostics is collected when either the environment variable
-		// AGENT_COLLECT_DIAG=true or the test is marked failed
-		collect := collectDiagFlag()
-		failed := f.t.Failed()
-		if collect || failed {
-			if collect {
-				f.t.Logf("collecting diagnostics; AGENT_COLLECT_DIAG=true")
-			} else if failed {
-				f.t.Logf("collecting diagnostics; test failed")
-			}
-			f.collectDiagnostics()
+	return out, nil
+}
+
+func (f *Fixture) cleanUpUninstall() bool {
+	if !f.installed {
+		f.t.Logf("skipping uninstall; agent not installed (fixture.installed is false)")
+		// not installed; no need to clean up or collect diagnostics
+		return true
+	}
+
+	// uninstall agent
+	// 5 minute timeout, to ensure that it at least doesn't get stuck.
+	// original context is not used as it could have a timeout on the context
+	// for the install and we don't want that context to prevent the uninstall
+	uninstallCtx, uninstallCancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer uninstallCancel()
+	out, err := f.Uninstall(uninstallCtx, &UninstallOpts{Force: true, UninstallToken: f.uninstallToken})
+	f.setClient(nil)
+	if err != nil &&
+		(errors.Is(err, ErrNotInstalled) ||
+			strings.Contains(
+				err.Error(),
+				"elastic-agent: no such file or directory")) {
```

Comment:
Since we're already refactoring, please assign each condition to a variable, this is very hard to read.